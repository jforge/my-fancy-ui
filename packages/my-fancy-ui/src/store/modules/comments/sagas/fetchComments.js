// @flow

import type {CommentType} from './../types.js';

import {put, call, takeLatest} from 'redux-saga/effects';
import {createEnvironmentSpecificSaga} from 'redux-lumbergh';
import {
  actions as commentsActions,
  actionTypes as commentsActionTypes
} from './../actions.js';

export function* fetchComments(): Generator<any, any, any> {
  const response = yield call(
    fetch,
    'https://jsonplaceholder.typicode.com/comments'
  );
  const comments: Array<CommentType> = yield response.json();

  yield put(commentsActions.setComments(comments));
}

function* watchFetchComments(): Generator<any, any, any> {
  yield takeLatest(commentsActionTypes.LOAD_COMMENTS, fetchComments);
}

export default [
  createEnvironmentSpecificSaga.server(fetchComments),
  createEnvironmentSpecificSaga.client(watchFetchComments)
];
