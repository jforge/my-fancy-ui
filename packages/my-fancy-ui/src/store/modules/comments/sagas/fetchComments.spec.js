// @flow

import {expectSaga} from 'redux-saga-test-plan';
import {fetchComments} from './fetchComments.js';

describe('fetchComments()', () => {
  let res;

  beforeEach(() => {
    res = {
      ok: true,
      headers: {
        get: () => 'application/json'
      },
      json: () => Promise.resolve(['someComments'])
    };
    global.fetch = jest.fn(() => Promise.resolve(res));
  });

  afterEach(() => {
    res = null;
    global.fetch = undefined;
  });

  it('should be a generator function.', () => {
    expect(typeof fetchComments).toBe('function');
    expect(typeof fetchComments().next).toBe('function');
  });

  it('should fetch the comments from the API and dispatch a "setComments" action with the response to the store.', async function() {
    const result = await expectSaga(fetchComments).run();

    expect(result.toJSON()).toMatchSnapshot();
  });
});
