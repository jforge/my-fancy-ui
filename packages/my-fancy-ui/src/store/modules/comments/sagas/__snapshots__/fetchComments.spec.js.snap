// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchComments() should fetch the comments from the API and dispatch a "setComments" action with the response to the store. 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "https://jsonplaceholder.typicode.com/comments",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "comments": Array [
              "someComments",
            ],
          },
          "type": "@@immowelt/my-fancy-ui/comments/SET_COMMENTS",
        },
        "channel": null,
      },
    },
  ],
}
`;
